### PR TITLE
Mildestone 4: Add Basic UI

### PR DESCRIPTION
**Pull request by digia3d**
In this part of the project I did the following:

> Create an app.rb file with methods that do the following:

- List all books.
- List all people.
- Create a person (teacher or student, not a plain Person).
- Create a book.
- Create a rental.
- List all rentals for a given person id.

> Create main.rb with a defined entry point, there is a method called main that is invoked at the end of the file.

- Present the user with a list of options to perform.
- Lets users choose an option.
- If needed, ask for parameters for the option.
- Have a way to quit the app.

### CODE REVIEW

File: student.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 require_relative './person'
 
 class Student < Person
-  def initialize(_classroom, age, name = 'Unknown', parent_permission: true)
+  def initialize(classroom, age, name = 'Unknown', parent_permission: true)
     super(age, name, parent_permission)
     @classroom = classroom
```
Comment (KDlamini) [inline student.rb]: When I create a student I get the following error.

image

- This is due to the arguments passed in the `Student` class. Kindly look into it.

File: app.rb [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    print 'Specialization: '
+    specialization = gets.chomp
+
+    teacher = Teacher.new(specialization, age, name)
+    @people << teacher
+
+    puts 'Teacher created successfully'
+    sleep 0.65
+    menu
+  end
```
Comment (KDlamini) [inline app.rb:85]: Same thing with creating a teacher. Kindly look into it and make sure that a teacher can be successfully created.

image

File: app.rb [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+def create_a_rental
+  puts 'Select a book from the following list by number'
+  @books.each_with_index { |book, index| puts "#{index} Title: #{book.title}, Author: #{book.author}" }
+  book_id = gets.chomp.to_i
+
+  puts 'Select a person from the following list by number (not id)'
+  @people.each_with_index do |person, index|
+    puts "#{index} [#{person.class}] Name: #{person.name}, ID: #{person.id}, Age: #{person.age}"
+  end
+  person_id = gets.chomp.to_i
```
Comment (KDlamini) [inline app.rb]: NOTE: When the issues are resolved, kindly ensure that all other functionalities associated with `students` and `teachers` work as expected.

- All created students and teachers can be listed successfully
- Rentals can be created using the list of persons (student or teacher) and books
- Rentals can be viewed using the ID of a person

**Review by KDlamini — CHANGES_REQUESTED** [review]
### Hi @digia3d 🙋‍♀️,  

Great work so far! 💯 
There are some issues that you still need to work on to improve your project according to the requirements. You are almost there! 💪🏾 

pugReaching

## Required Changes ♻️
 
**Kindly look through the comments below 👇 for required changes**

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app.rb [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+def list_rentals_by_person_id
+  print 'ID of person: '
+  id = gets.chomp.to_i
+  puts 'Rentals:'
+  @rentals.each do |rental|
+    puts "Date: #{rental.date}, Book: '#{rental.book.title}' by #{rental.book.author}" if rental.person.id == id
+  end
+  sleep 0.65
+  menu
+end
```
Comment (cynthiainga) [inline app.rb]: - [x] Good job by implementing the changes required but there are still some which aren't been addressed yet. As required by the previous reviewer you should ensure that all other functionalities associated with students and teachers work as expected 🙃 🙃 
- [x] When I try to list rentals by person id I get an error `undefined method id`, so please kindly fix that 😉 

Screen Shot 2022-09-30 at 09 49 01

**Review by cynthiainga — CHANGES_REQUESTED** [review]
Hi @digia3d, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.



 ## Highlights

- [ ] No linters errors ✔️ 
- [ ] Descriptive PR ✔️ 
- [ ] Good commit messages ✔️ 
- [ ] Used correct Gitflow ✔️ 

 ## Required Changes ♻️

- [x] Please kindly check the comment under the review 👇 👇 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app.rb [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  puts 'Select a person from the following list by ID'
+  @people.each { |person| puts "ID: #{person.id} Name: #{person.name} Age:#{person.age}" }
+  id = gets.chomp
+  puts 'Rentals: '
+  @rentals.each do |rental|
+    puts "Date: #{rental.date} - Book: #{rental.book.title} - Author: #{rental.book.author}" if rental.person.id == id
+  end
+  sleep 0.65
+  menu
+end
```
Comment (thecodechaser) [inline app.rb]: - [x] As requested by the previous reviewer, All functionalities of the application should work as expected. When I try to list rentals for a person I see nothing on the screen. Please fix this issue and make sure list rentals work as expected.
image

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @digia3d👋, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## To Highlight 🎉
- No linter errors. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by m05e5 — APPROVED** [review]
# Status: ✔️ Project Approved ✔️ 

Hello @digia3d    🔥 

Good job on implementing the reqested changes. You smashed it. Well Done.
Your project is complete! There is nothing else to say other than... it's time to merge it :shipit:
Congratulations! 🎉

goodjob


Now you can merge this branch 🚀 

Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

_m05e5_

### 💻 👩🏽‍💻  Happy Coding!  👨🏽‍💻 🖥️

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**